prover/src/matrix/row_matrix.rs

```diff
@@ -0,0 +1,155 @@
+use crate::{matrix::ARR_SIZE, Matrix};
+use math::{fft, FieldElement, StarkField};
+use utils::collections::Vec;
+use utils::{flatten_vector_elements, uninit_vector};
+
+// ROWMAJOR MATRIX
+// ================================================================================================
+
+/// A row-major matrix of field elements. The matrix is represented as a single vector of field
+/// elements, where the first `row_len` elements represent the first row of the matrix, the next
+/// `row_len` elements represent the second row, and so on.
+///
+/// # Note
+/// - The number of rows in the matrix is always a multiple of the ARR_SIZE.
+/// - The number of columns in the matrix is always a multiple of the ARR_SIZE.
+#[derive(Clone, Debug)]
+pub struct RowMatrix<E: FieldElement> {
+    data: Vec<E>,
+    row_len: usize,
+}
+
+impl<E> RowMatrix<E>
+where
+    E: FieldElement,
+{
+    // CONSTRUCTORS
+    // --------------------------------------------------------------------------------------------
+
+    /// Creates a new row-major matrix from the specified data and row length. The data must be
+    /// arranged in row-major order, i.e. the first `row_len` elements of the data represent the first
+    /// row of the matrix, the next `row_len` elements represent the second row, and so on.
+    ///
+    /// # Panics
+    /// - if the number of elements in the data is not a multiple of the specified row length;
+    /// - if the specified row length is not a multiple of the ARR_SIZE;
+    /// - if the specified data is empty.
+    pub fn new(data: Vec<E>, row_len: usize) -> Self {
+        assert!(data.len() % row_len == 0);
+        assert!(row_len % ARR_SIZE == 0);
+        assert!(!data.is_empty());
+
+        Self { data, row_len }
+    }
+
+    /// Converts a column-major matrix of polynomials `Matrix<E>` into a RowMatrix evaluated at
+    /// a shifted domain offset.
+    pub fn from_polys(polys: &Matrix<E>, blowup_factor: usize) -> Self {
+        // get the number of rows and columns in the polys.
+        let row_width = polys.num_cols();
+        let num_rows = polys.num_rows();
+
+        // get the twiddles for the segment.
+        let twiddles = fft::get_twiddles::<E::BaseField>(polys.num_rows() * blowup_factor);
+        let domain_offset = E::BaseField::GENERATOR;
+
+        // get the number of segments in the matrix.
+        let num_of_segments = row_width / ARR_SIZE;
+        let mut segments = Segments::new(Vec::new());
+
+        // precompute offsets for each row.
+        let mut offsets = Vec::with_capacity(num_rows);
+        offsets.push(E::BaseField::ONE);
+        for i in 1..num_rows {
+            offsets.push(offsets[i - 1] * domain_offset);
+        }
+
+        // create segments.
+        (0..num_of_segments).for_each(|i| {
+            // create a vector of arrays to hold the result.
+            let mut result_vec_of_arrays =
+                unsafe { uninit_vector::<[E; ARR_SIZE]>(num_rows * blowup_factor) };
```

Comment:
I wonder if it would be more efficient to create a single vector of size `num_rows * row_width * blowup_factor` and then break it into individual vectors via [from_raw_parts()](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.from_raw_parts).

Either way, I'd probably encapsulate this memory allocation logic into a stand-alone function which is run before we enter this loop. The function could have the following signature:
```Rust
fn allocate_segments(num_rows: usize, row_width: usize, blowup_factor: usize) -> Vec<Vec<[E; ARR_SIZE]>>
```

Reply:
I think it get bit easier if we just return a Vec of `Segment` instead of `unallocated_vector`. We are using the output in `prepare_segment` which should return a `Segment`. Given this vector is not made up of  `Segment`, we might not be able to update the input vector.